src/librustc_mir/interpret/validity.rs

```diff
@@ -262,8 +264,12 @@ impl<'rt, 'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>>
         variant_id: VariantIdx,
         new_op: OpTy<'tcx, M::PointerTag>
     ) -> EvalResult<'tcx> {
-        let name = old_op.layout.ty.ty_adt_def().unwrap().variants[variant_id].ident.name;
-        self.visit_elem(new_op, PathElem::Variant(name))
+        let name = match old_op.layout.ty.ty_adt_def() {
+            Some(def) => PathElem::Variant(def.variants[variant_id].ident.name),
+            // Generators also have variants but no def
```

Comment:
Maybe pattern-match on the type explicitly?

Reply:
Done.